### ISSUE TITLE
Facenet on Raspberry Pi with MYRIAD

### ISSUE THREAD

**Issue by bsekachev** [opencv/dldt#246]
Hello, 
I want to infer [FaceNet](https://github.com/davidsandberg/facenet/wiki/Training-using-the-VGGFace2-dataset) model on MYRIAD (Raspberry Pi 3B+ as a host).

1. I have downloaded the model [here](https://drive.google.com/file/d/0B5MzpY9kBtDVZ2RpVDYwWmxoSUk)
2. I have converted the model with model optimizer (from OpenVINO 2019R2) 
Conversion command is: 
```
python3 mo.py --input_model=path/to/pb --output_dir=/out/dir --data_type=FP16 --freeze_placeholder_with_value "phace_train->False"
```
3. Now I can infer converted on Raspberry. Lucky? No. 

Actually now it seems to be, that after I uploaded model to myriad plugin, memory is corrupted. 
So, I get segfault errors with different stacktraces at random time even after a lot of successful inferences and even if I don't run an inference at all.  

- On raspberry I use OpenVINO from [this](https://download.01.org/opencv/2019/openvinotoolkit/R2/).
- This model worked on CPU (I tried FP32) plugin without any problems. 
- Code I use to create an executable model:
```cpp
IEFacenet_V1::IEFacenet_V1(const std::string xml, const std::string bin, const std::string device) {
    using namespace InferenceEngine; 

    Core ie;
    // Reading a network
    CNNNetReader networkReader;
    networkReader.ReadNetwork(xml);
    networkReader.ReadWeights(bin);

    this->_network = networkReader.getNetwork();
    this->_network.setBatchSize(1);

    // Get information about topology
    InputsDataMap inputInfo(this->_network.getInputsInfo());
    OutputsDataMap outputInfo(this->_network.getOutputsInfo());

    this->_executable = ie.LoadNetwork(this->_network, device);
    this->_infer_request = this->_executable.CreateInferRequest();
    this->_input = this->_infer_request.GetBlob((*inputInfo.begin()).first);
    this->_output = this->_infer_request.GetBlob((*outputInfo.begin()).first);
};
```

If I don't execute this code, the program works well with fake data. 

One of stacktraces for example:
```
Thread 1 "CPPExample" received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) backtrace
#0  0x75b32f24 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x75b1e230 in __GI_abort () at abort.c:79
#2  0x75b6e51c in __libc_message (action=action@entry=do_abort, fmt=<optimized out>) at ../sysdeps/posix/libc_fatal.c:181
#3  0x75b75044 in malloc_printerr (str=<optimized out>) at malloc.c:5341
#4  0x75b752fc in malloc_consolidate (av=av@entry=0x75c517d4 <main_arena>) at malloc.c:4488
#5  0x75b78694 in _int_malloc (av=av@entry=0x75c517d4 <main_arena>, bytes=bytes@entry=32600) at malloc.c:3695
#6  0x75b7a318 in __GI___libc_malloc (bytes=32600) at malloc.c:3057
#7  0x75d76f60 in operator new(unsigned int) () at /usr/lib/arm-linux-gnueabihf/libstdc++.so.6
#8  0x6f310bd8 in  () at /opt/intel/openvino/deployment_tools/inference_engine/lib/armv7l/libmyriadPlugin.so
#9  0x6f30f724 in  () at /opt/intel/openvino/deployment_tools/inference_engine/lib/armv7l/libmyriadPlugin.so
#10 0x6f3095a2 in  () at /opt/intel/openvino/deployment_tools/inference_engine/lib/armv7l/libmyriadPlugin.so
#11 0x6f2f477c in  () at /opt/intel/openvino/deployment_tools/inference_engine/lib/armv7l/libmyriadPlugin.so
#12 0x76f9a2b8 in InferenceEngine::ExecutableNetwork::CreateInferRequest() () at libCPPClassificator.so
#13 0x76f960f4 in IEFacenet_V1::IEFacenet_V1(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) () at libCPPClassificator.so
```

**Comment by shubha-ramani** [comment]
Dear @bsekachev 

Sorry you're experiencing this technical glitch !  May I ask one simple troubleshooting technique though  ? I'm trying to remove Raspberry PI from the equation. do you happen to have a Windows or Linux computer that you can stick your NCS2 into and try again (after installing OpenVino of course) ? The reason is, I'm not convinced it's RPI. Maybe it's just Myriad. 

Let me know if it's possible for you to try this experiment.

Shubha

**Comment by bsekachev** [comment]
@shubha-ramani 
Thanks for answer. 

Actually I don't think that Raspberry PI is a reason. 
I am going to setup Ubuntu 18.04 in virtual box and run it there. 
I'll update this issue ASAP.

**Comment by shubha-ramani** [comment]
Dear @bsekachev 
Thank you for your prompt response ! Looking at your errors above I don't think Raspberry PI is the reason either. But thank you kindly for verifying !

Shubha

**Comment by bsekachev** [comment]
@shubha-ramani 
Another issue with virtual box. 

Can not initialize the stick. I get error message:
```
E: [ncAPI] [    721949] [CPPExample] ncDeviceOpen:844	Failed connection to device (1.3-ma2480) with error 7
terminate called after throwing an instance of 'InferenceEngine::details::InferenceEngineException'
  what():  Can not init Myriad device: NC_ERROR
Aborted (core dumped)
```

Then stick automatically disconnects (I haven't come across such behaviour before). I need to remove stic and insert it again (otherwise virtual box doesn't show stick within usb devices). Maybe some issues with using stick under a virtual machine. 

``dmesg`` output after connecting:
```
[  435.329668] usb 1-3: USB disconnect, device number 5
[  439.184903] usb 1-3: new high-speed USB device number 6 using xhci_hcd
[  439.561601] usb 1-3: New USB device found, idVendor=03e7, idProduct=2485, bcdDevice= 0.01
[  439.561602] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  439.561603] usb 1-3: Product: Movidius MyriadX
[  439.561604] usb 1-3: Manufacturer: Movidius Ltd.
[  439.561605] usb 1-3: SerialNumber: 03e72485
```

`` cat /etc/udev/rules.d/97-myriad-usbboot.rules `` output:
```
SUBSYSTEM=="usb", ATTRS{idProduct}=="2150", ATTRS{idVendor}=="03e7", GROUP="users", MODE="0660", ENV{ID_MM_DEVICE_IGNORE}="1"
SUBSYSTEM=="usb", ATTRS{idProduct}=="2485", ATTRS{idVendor}=="03e7", GROUP="users", MODE="0660", ENV{ID_MM_DEVICE_IGNORE}="1"
SUBSYSTEM=="usb", ATTRS{idProduct}=="f63b", ATTRS{idVendor}=="03e7", GROUP="users", MODE="0660", ENV{ID_MM_DEVICE_IGNORE}="1"
```

``lsusb`` output:
```
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 006: ID 03e7:2485  
Bus 001 Device 003: ID 80ee:0030 VirtualBox 
Bus 001 Device 002: ID 80ee:0021 VirtualBox USB Tablet
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

OpenVINO: 2019.2.275

Also a short question. This openvino version has two files: ``MvNCAPI-ma2450.mvcmd`` and ``MvNCAPI-ma2x8x.mvcmd`` in ``<openvino_dir>/inference_engine/lib/arch/``. The first is for NCS v1, the second for NCS2. But why the second has such a name? Should I rename the second to ``MvNCAPI-ma2485.mvcmd``. I put these files into `` /usr/local/lib/mvnc/ `` for now (found such possible solution in google, but it doesn't work).

**Comment by shubha-ramani** [comment]
Dear @bsekachev 
It seems like you've run into several issues here, one being with VirtualBox.  Please review [this IDZ Post](https://software.intel.com/en-us/forums/computer-vision/topic/805215) to see if it helps you.

And I am not understanding your question here ? As a general rule you probably shouldn't be messing with/renaming those files which are system files within the installation.

> Also a short question. This openvino version has two files: MvNCAPI-ma2450.mvcmd and MvNCAPI-ma2x8x.mvcmd in <openvino_dir>/inference_engine/lib/arch/. The first is for NCS v1, the second for NCS2. But why the second has such a name? Should I rename the second to MvNCAPI-ma2485.mvcmd. I put these files into /usr/local/lib/mvnc/ for now (found such possible solution in google, but it doesn't work).

Also did you follow the [Model Optimizer FaceNet Instructions](http://docs.openvinotoolkit.org/latest/_docs_MO_DG_prepare_model_convert_model_tf_specific_Convert_FaceNet_From_Tensorflow.html) carefully ? Please don't forget to add --data_type FP16 for MYRIAD.

Let me know. If you followed the above instructions carefully and added --data_type FP16, yet you're still experiencing the problem, then I shall try myself using MYRIAD plugged into my plain old Windows Host. 

Normally with OpenVino Inference Engine and Model Optimizer are not platform/host specific - at least such problems are extremely rare. The code was written in such a way to abstract out the platform, OS or host.

Shubha

**Comment by bsekachev** [comment]
> Also did you follow the Model Optimizer FaceNet Instructions carefully ? Please don't forget to add --data_type FP16 for MYRIAD.

Yes, I did. I have attached the command which I used for model optimizer in the first message. 

> Please review this IDZ Post to see if it helps you.

It seems to be, that the advice of Nayanbuu, Purevdorj helped me. 
Now I don't get any segfaults and program always closes correctly (on raspberry it absolutely always crashes when the thread of execution left the function main()).  

So, what can be a reason? Different OpenVINO packages? 
Rasbperry: **2019.2.242**
VirtualBox: **2019.2.275**

**Comment by shubha-ramani** [comment]
@bsekachev 
Great data point. So I guess I was wrong - this may very well be related to the RPI environment.  So using Ubuntu on VirtualBox your code runs to completion without problems but on RPI it bombs.  Same exact inference code, same exact model.  Am I correct ? If so, this is enough evidence to file a bug.

And thank you kindly for doing this additional test for me. It helps a lot.

Shubha

**Comment by bsekachev** [comment]
@shubha-ramani 

Yes. Exactly the same code and the same model.

**Comment by shubha-ramani** [comment]
Dear @bsekachev 

Thank you for your patience. I will file a bug promptly !

Shubha

**Comment by bsekachev** [comment]
@shubha-ramani 

I am going to do native compilation of dldt from sources on raspberry. 
Do you think it can help?

**Comment by shubha-ramani** [comment]
@bsekachev 
Certainly it's worth a shot. If it were the Myriad Plugin it would have failed also on Ubuntu. I'm curious to know the result.

So, sure - try it. Let me know if it solves your problem !

Shubha

**Comment by bsekachev** [comment]
@shubha-ramani 

Oh, finally I have found what was the problem. 
When I read how to do crosscompilation, I come across the note:
> NOTE: Native applications that link to cross-compiled Inference Engine library require an extra compilation flag -march=armv7-a.

In tracebacks I had seen Jenkins paths, so I decided that this OpenVINO package wasn't compiled natively. 
After I compiled my app with this flag on raspberry, everything works fine. 
To sum up, I believe OpenVINO team should do a little more emphasis on this detail in [getting started](https://docs.openvinotoolkit.org/latest/_docs_install_guides_installing_openvino_raspbian.html) :)

There is this flag in a command for build a sample, but actually it isn't very noticeable (at least for me). 

Finally, I suppose I can close this issue. 
Thanks for help
